src/Core/src/Credentials/ContainerProvider.php

```diff
@@ -77,4 +100,78 @@ public function getCredentials(Configuration $configuration): ?Credentials
             Credentials::adjustExpireDate(new \DateTimeImmutable($result['Expiration']), $date)
         );
     }
+
+    /**
+     * Checks if the provided IP address is a loopback address.
+     *
+     * @param string $host the host address to check
+     *
+     * @return bool true if the IP is a loopback address, false otherwise
+     */
+    public static function isLoopBackAddress(string $host)
+    {
+        // Validate that the input is a valid IP address
+        if (!filter_var($host, \FILTER_VALIDATE_IP)) {
+            return false;
+        }
+
+        // Convert the IP address to binary format
+        $packedIp = inet_pton($host);
+
+        // Check if the IP is in the 127.0.0.0/8 range
+        if (4 === \strlen($packedIp)) {
+            return 127 === \ord($packedIp[0]);
+        }
+
+        // Check if the IP is ::1
+        if (16 === \strlen($packedIp)) {
+            return $packedIp === inet_pton('::1');
+        }
+
+        // Unknown IP format
+        return false;
+    }
+
+    private function getFullUri(Configuration $configuration): ?string
+    {
+        $relativeUri = $configuration->get(Configuration::OPTION_CONTAINER_CREDENTIALS_RELATIVE_URI);
+        $fullUri = $configuration->get(Configuration::OPTION_POD_IDENTITY_CREDENTIALS_FULL_URI);
+
+        if (null !== $relativeUri) {
+            return self::ECS_ENDPOINT . $relativeUri;
+        }
+
+        return $fullUri;
+    }
+
+    private function getHeaders(?string $tokenFileContent): array
+    {
+        return $tokenFileContent ? ['Authorization' => $tokenFileContent] : [];
+    }
+
+    private function isUriValid(string $uri): bool
+    {
+        $parsedUri = parse_url($uri);
+        if (false === $parsedUri) {
+            return false;
+        }
+
+        if (!isset($parsedUri['scheme'])) {
+            return false;
+        }
+
+        if ('https' !== $parsedUri['scheme']) {
+            $host = trim($parsedUri['host'] ?? '', '[]');
+
+            $ecsHost = parse_url(self::ECS_ENDPOINT)['host'] ?? '';
+            $eksHost = self::EKS_HOST_IPV4;
+
+            // check if the host is a valid ecs or eks host or the loopback address
+            if ($host !== $ecsHost && $host !== $eksHost && self::EKS_HOST_IPV6 !== $host && !self::isLoopBackAddress($host)) {
```

Comment:
feel inconsistent to store ipv4 in a variable, but not for ipv6 :thinking: 

```suggestion
            $eksHost = self::EKS_HOST_IPV4;

            // check if the host is a valid ecs or eks host or the loopback address
            if ($host !== $ecsHost && $host !== self::EKS_HOST_IPV4 && self::EKS_HOST_IPV6 !== $host && !self::isLoopBackAddress($host)) {
```


also, for performance reason, we can re-order logic to check easy computation first

```php
if ('https' !== $parsedUri['scheme']) {
  $host = trim($parsedUri['host'] ?? '', '[]');
  if (self::EKS_HOST_IPV4 === $host || self::EKS_HOST_IPV6 === $host) {
    return true;
  }

  if ($host === parse_url(self::ECS_ENDPOINT, \PHP_URL_HOST)) {
    return true;
  }

  return self::isLoopBackAddress($host);
}
```


also, if we store `ECS_HOST = 169.254.170.2` (instead of endpoint) we could save the `parse_url` thing without compromising too much the readability